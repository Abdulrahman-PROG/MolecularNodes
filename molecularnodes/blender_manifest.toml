schema_version = "1.0.0"

# Example of manifest file for a Blender extension
# Change the values according to your extension
id = "molecularnodes"
version = "4.2.0"
name = "molecularnodes"
tagline = "Import and animation of molecular data."
maintainer = "Brady Johnston <brady.johnston@me.com>"
type = "add-on"

# Optional: add-ons can list which resources they will require:
# * "files" (for access of any filesystem operations)
# * "network" (for internet access)
# * "clipboard" (to read and/or write the system clipboard)
# * "camera" (to capture photos and videos)
# * "microphone" (to capture audio)
permissions = [
	"files",
	"network",
]

# Optional link to documentation, support, source files, etc
website = "http://bradyajohnston.github.io/MolecularNodes"

# Optional list defined by Blender and server, see:
# https://docs.blender.org/manual/en/dev/extensions/tags.html
tags = [
	"3D View",
	"Animation",
	"Node",
]

blender_version_min = "4.2.0"
# Optional: maximum supported Blender version
# blender_version_max = "5.1.0"

# License conforming to https://spdx.org/licenses/ (use "SPDX: prefix)
# https://docs.blender.org/manual/en/dev/extensions/licenses.html
license = [
  "SPDX:GPL-2.0-or-later",
]
# Optional: required by some licenses.
# copyright = [
#   "2002-2024 Developer Name",
#   "1998 Company Name",
# ]

# Optional: bundle 3rd party Python modules.
wheels = [
	"./wheels/pandas-2.2.2-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/urllib3-2.2.1-py3-none-any.whl",
	"./wheels/python_dateutil-2.9.0.post0-py2.py3-none-any.whl",
	"./wheels/starfile-0.5.6-py3-none-any.whl",
	"./wheels/typing_extensions-4.11.0-py3-none-any.whl",
	"./wheels/pyarrow-16.0.0-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/numpy-1.26.4-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/biotite-0.40.0-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/requests-2.31.0-py3-none-any.whl",
	"./wheels/pytz-2024.1-py2.py3-none-any.whl",
	"./wheels/msgpack-1.0.8-cp311-cp311-macosx_11_0_arm64.whl",
	"./wheels/tzdata-2024.1-py2.py3-none-any.whl",
	"./wheels/idna-3.7-py3-none-any.whl",
	"./wheels/six-1.16.0-py2.py3-none-any.whl",
	"./wheels/mrcfile-1.4.3-py2.py3-none-any.whl",
	"./wheels/certifi-2024.2.2-py3-none-any.whl",
	"./wheels/networkx-3.3-py3-none-any.whl",
	"./wheels/charset_normalizer-3.3.2-cp311-cp311-macosx_11_0_arm64.whl",
]
