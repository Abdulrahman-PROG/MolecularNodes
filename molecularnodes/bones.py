import bpy
import numpy as np
from . import obj

def add_bones(object):
    ## creates bones and assigns correct weights
    
    bone_positions = get_bone_positions(object)
    # weights = get_bone_weights()
    
    bones = create_bones(bone_positions)
    # assign_bone_weights(object, bones, weights)
    
    return bones

def get_bone_positions(object):
    
    positions, atom_name, chain_id, res_id, sec_struct = [
        obj.get_attribute(object, att) 
        for att in ['position', 'atom_name', 'chain_id', 'res_id', 'sec_struct']
        ]
    
    sec_struct_change = np.where(np.diff(sec_struct))[0]

    pos_idx = np.where(sec_struct_change)[0]
    bone_groups = np.cumsum(sec_struct_change)
    # bone_pos_a = positions[pos_idx - 1 , :]
    bone_positions = positions[ sec_struct_change, :]
    
    
    
    
    
    print("get_bone_positions")
    return bone_positions

def get_bone_weights(object):
    print("get bone weights")

def create_bones(positions):
    
    bpy.ops.object.add(type='ARMATURE', enter_editmode=True)
    object = bpy.context.object
    object.name = 'mn_armature'
    armature = object.data
    armature.name = 'mn_frame'
    arm_name = armature.name
    bones = []
    # add bones
    for i, position in enumerate(positions):
        if i >=  len(positions) + 1:
            continue
        bone_name = f"mn_armature_{i}"
        bone = armature.edit_bones.new(bone_name)
        bone.head = position
        try:
            bone.tail = positions[i + 1, :]
        except:
            pass
        bones.append(bone.name)
    
    bpy.ops.object.editmode_toggle()
    bpy.ops.object.editmode_toggle()

    armature = bpy.data.armatures[arm_name]
    bones_a = bones.copy()
    bones_b = bones.copy()
    bones_b.pop(0)
    bones = zip(bones_a, bones_b)
    # print(f"{list(bones)=}")
    for bone_a, bone_b in bones:
        print(f"{bone_a=}")
        print(f"{bone_b=}")
        armature.edit_bones.active = armature.edit_bones[bone_a]
        for bone in [bone_a, bone_b]:
            armature.edit_bones[bone].select = True
        bpy.ops.armature.parent_set(type='CONNECTED')
        for bone in [bone_a, bone_b]:
            armature.edit_bones[bone].select = False

    print("create boens")

def assign_bone_weights(object, bones, weights):
    print("assigning bone weights")
