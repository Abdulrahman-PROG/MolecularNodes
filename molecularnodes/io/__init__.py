from .parse import (
    CIF,
    BCIF,
    PDB,
    SDF,
    CellPack,
    StarFile,
    MMTF,
    MDAnalysisSession
)
from .wwpdb import fetch
from .local import load
