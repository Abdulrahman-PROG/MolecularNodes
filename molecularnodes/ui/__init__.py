from .ops import ops_ui
