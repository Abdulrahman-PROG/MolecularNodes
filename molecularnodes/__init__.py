# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTIBILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

import bpy

from .io import ops_io
from .io.parse.mda import _rejuvenate_universe, _sync_universe
from .io.parse.star import _rehydrate_ensembles
from .props import MolecularNodesObjectProperties
from .ui.node_menu import MN_add_node_menu, draw_node_menus
from .ui.ops import ops_ui
from .ui.panel import MN_PT_panel, change_style_menu, change_style_node_menu
from .utils import MN_OT_Install_Template

all_classes = ops_ui + ops_io + [MN_OT_Install_Template, MolecularNodesObjectProperties]

universe_funcs = [_sync_universe, _rejuvenate_universe]


def register():
    bpy.utils.register_class(MN_PT_panel)
    bpy.types.NODE_MT_add.append(MN_add_node_menu)

    # register all of the import operators
    for op in all_classes:
        bpy.utils.register_class(op)
    bpy.types.Object.mn = bpy.props.PointerProperty(type=MolecularNodesObjectProperties)
    bpy.types.Object.mn = bpy.props.PointerProperty(type=MolecularNodesObjectProperties)
    bpy.types.VIEW3D_MT_object_context_menu.prepend(change_style_menu)
    bpy.types.NODE_MT_context_menu.prepend(change_style_node_menu)
    for func in universe_funcs:
        try:
            bpy.app.handlers.load_post.append(func)
        except ValueError as e:
            print(f"Filaed to append {func}, error: {e}.")


def unregister():
    bpy.utils.unregister_class(MN_PT_panel)
    bpy.types.NODE_MT_add.remove(MN_add_node_menu)

    # unregister all of the import operator classes
    for op in all_classes:
        bpy.utils.unregister_class(op)

    try:
        bpy.types.NODE_MT_add.remove(MN_add_node_menu)
        bpy.types.VIEW3D_MT_object_context_menu.remove(change_style_menu)
        bpy.types.NODE_MT_context_menu.remove(change_style_node_menu)
        bpy.app.handlers.load_post.append(_rehydrate_ensembles)

        del bpy.types.Object.mn
        for func in universe_funcs:
            try:
                bpy.app.handlers.load_post.remove(func)
            except ValueError as e:
                print(f"Failed to remove {func}, error: {e}.")
    except RuntimeError:
        pass


# # can't register the add-on when these are uncommnted, but they do fix the issue
# # of having to call register() when running a script
# # unregister()
# # register()

# # # register won't be called when MN is run as a module
# bpy.app.handlers.load_post.append(_rejuvenate_universe)
# bpy.app.handlers.load_post.append(_rehydrate_ensembles)
# bpy.app.handlers.save_post.append(_sync_universe)
