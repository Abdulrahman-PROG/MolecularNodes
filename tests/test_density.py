import molecularnodes as mn
import bpy
import pytest
import numpy as np
from .constants import test_data_directory
from .utils import (
    apply_mods
)
try:
    import pyopenvdb
except ImportError:
    pytest.skip("pyopenvdb not installed", allow_module_level=True)

mn.unregister()
mn.register()

@pytest.fixture
def density_file():
    file = test_data_directory / "emd_24805.map.gz"
    vdb_file = test_data_directory / "emd_24805.vdb"
    vdb_file.unlink(missing_ok=True)
    # Make all densities are removed
    for o in bpy.data.objects:
        if o.mn.molecule_type == "density":
            bpy.data.objects.remove(o,do_unlink=True)
    return file

def test_density_load(density_file):
    
    obj = mn.io.density.load(density_file,style="density_surface")
    evaluated = mn.blender.obj.evaluate_using_mesh(obj)
    pos = mn.blender.obj.get_attribute(evaluated,"position")

    assert len(pos) > 1000

    avg = np.mean(pos,axis=0)
    assert np.linalg.norm(avg) > 0.5 

    assert obj.mn.molecule_type == "density"
    assert obj.users_collection[0] == mn.blender.coll.mn()

def test_density_centered(density_file):

    # First load using standar parameters to test recreation of vdb
    o = mn.io.density.load(density_file,style="density_surface")
    # Then refresh the scene
    bpy.data.objects.remove(o,do_unlink=True)

    obj = mn.io.density.load(density_file,style="density_surface",center=True)
    evaluated = mn.blender.obj.evaluate_using_mesh(obj)

    pos = mn.blender.obj.get_attribute(evaluated,"position")

    assert len(pos) > 1000

    avg = np.mean(pos,axis=0)
    assert np.linalg.norm(avg) < 0.1

def test_density_invert(density_file):

    # First load using standar parameters to test recreation of vdb
    o = mn.io.density.load(density_file,style="density_surface")
    # Then refresh the scene
    bpy.data.objects.remove(o,do_unlink=True)

    obj = mn.io.density.load(density_file,style="density_surface",invert=True)
    style_node = mn.blender.nodes.get_style_node(obj)
    style_node.inputs["Threshold"].default_value = 0.01
    evaluated = mn.blender.obj.evaluate_using_mesh(obj)

    pos = mn.blender.obj.get_attribute(evaluated,"position")    
    # At this threshold after inverting we should have a cube the size of the volume
    assert pos[:,0].max() > 2.0
    assert pos[:,1].max() > 2.0
    assert pos[:,2].max() > 2.0

def test_density_multiple_load():
    file = test_data_directory / "emd_24805.map.gz"
    obj = mn.io.density.load(file,style="density_surface")
    obj2 = mn.io.density.load(file,style="density_surface")

    assert obj.mn.molecule_type == "density"
    assert obj2.mn.molecule_type == "density"
    assert obj.users_collection[0] == mn.blender.coll.mn()
    assert obj2.users_collection[0] == mn.blender.coll.mn()