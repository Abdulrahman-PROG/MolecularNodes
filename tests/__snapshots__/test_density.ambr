# serializer version: 1
# name: test_density_operator[False-False-False]
  []
# ---
# name: test_density_operator[False-False-True]
  []
# ---
# name: test_density_operator[False-True-False]
  [[1.20 1.53 1.23]
   [0.89 0.96 1.46]
   [1.22 1.51 0.95]
   [1.32 0.71 0.99]
   [0.79 0.87 1.29]
   [0.90 0.80 0.96]
   [1.03 0.88 1.14]
   [1.32 1.28 1.38]
   [1.19 1.86 1.23]
   [1.16 1.58 0.95]
   [1.34 1.84 1.30]
   [1.16 0.94 1.39]
   [1.12 0.85 1.21]
   [1.06 0.69 1.04]
   [1.28 0.88 0.98]
   [1.21 1.62 1.03]
   [1.06 1.58 1.11]
   [1.39 1.54 1.42]
   [1.23 1.02 1.41]
   [1.09 1.44 1.09]
   [1.11 1.44 1.18]
   [1.29 1.61 1.31]
   [0.97 0.74 1.10]
   [1.03 1.22 1.28]
   [1.19 1.31 1.21]
   [1.19 1.20 1.10]
   [1.32 2.06 1.28]
   [1.38 1.56 1.35]
   [1.27 0.86 0.85]
   [1.27 1.36 1.09]
   [1.23 1.40 1.13]
   [1.30 1.74 1.15]
   [0.94 0.63 1.14]
   [1.27 0.88 1.28]
   [1.11 0.99 1.42]
   [0.99 0.73 1.29]
   [1.18 0.91 1.14]
   [1.24 1.02 1.10]
   [1.26 0.92 0.79]
   [1.36 1.01 1.07]
   [1.27 2.09 1.20]
   [0.95 0.77 1.08]
   [1.51 1.47 1.15]
   [1.07 1.50 1.15]
   [1.38 1.07 1.09]
   [1.25 1.85 1.24]
   [1.33 0.93 1.02]
   [0.96 0.76 1.27]
   [0.83 0.84 1.26]
   [1.18 1.96 1.18]
   [1.12 0.59 0.87]
   [1.28 1.26 1.26]
   [1.17 0.85 1.49]
   [1.22 1.56 0.92]
   [1.01 1.22 1.46]
   [1.36 1.50 1.06]
   [1.18 1.54 1.20]
   [1.24 2.09 1.17]
   [1.30 2.00 1.10]
   [1.27 1.56 0.97]
   [1.02 0.81 0.97]
   [1.22 1.80 1.19]
   [1.20 0.90 1.11]
   [1.25 0.74 0.94]
   [1.20 1.30 1.07]
   [1.31 1.00 1.16]
   [1.08 0.61 1.14]
   [1.18 0.84 1.29]
   [1.18 2.01 1.18]
   [1.00 1.03 1.30]
   [0.96 0.63 1.23]
   [1.28 1.44 1.07]
   [1.27 1.57 1.39]
   [1.08 1.16 1.17]
   [1.04 1.55 1.25]
   [0.97 0.95 1.45]
   [1.22 1.34 1.24]
   [1.32 1.28 1.38]
   [1.29 0.86 0.94]
   [1.08 0.76 1.24]
   [1.37 1.41 1.28]
   [1.05 1.44 1.16]
   [1.13 1.04 1.46]
   [1.21 1.45 1.06]
   [1.14 0.68 0.87]
   [1.05 0.76 1.43]
   [1.21 1.25 1.11]
   [1.10 0.78 0.85]
   [1.28 2.03 1.18]
   [1.08 1.12 1.35]
   [0.95 0.91 0.93]
   [1.05 1.14 1.14]
   [0.93 1.18 1.16]
   [1.05 0.65 0.91]
   [1.33 1.07 1.33]
   [1.10 0.98 1.29]
   [1.14 1.25 1.16]
   [1.27 1.47 1.15]
   [1.22 1.55 1.30]
   [1.24 1.57 1.16]]
# ---
# name: test_density_operator[False-True-True]
  [[ 0.00e+00  3.34e-01  3.48e-02]
   [-3.07e-01 -2.32e-01  2.66e-01]
   [ 2.53e-02  3.15e-01 -2.42e-01]
   [ 1.22e-01 -4.90e-01 -2.02e-01]
   [-4.02e-01 -3.30e-01  9.96e-02]
   [-2.95e-01 -3.95e-01 -2.37e-01]
   [-1.66e-01 -3.12e-01 -5.45e-02]
   [ 1.29e-01  8.30e-02  1.88e-01]
   [-1.84e-03  6.61e-01  3.16e-02]
   [-3.39e-02  3.81e-01 -2.47e-01]
   [ 1.49e-01  6.46e-01  1.06e-01]
   [-3.19e-02 -2.57e-01  1.93e-01]
   [-7.43e-02 -3.40e-01  1.66e-02]
   [-1.38e-01 -5.04e-01 -1.58e-01]
   [ 7.98e-02 -3.19e-01 -2.12e-01]
   [ 1.58e-02  4.23e-01 -1.66e-01]
   [-1.36e-01  3.87e-01 -8.04e-02]
   [ 1.96e-01  3.49e-01  2.29e-01]
   [ 3.87e-02 -1.79e-01  2.20e-01]
   [-1.00e-01  2.50e-01 -1.02e-01]
   [-8.92e-02  2.47e-01 -1.93e-02]
   [ 9.15e-02  4.15e-01  1.14e-01]
   [-2.28e-01 -4.52e-01 -9.55e-02]
   [-1.62e-01  2.10e-02  8.00e-02]
   [-6.39e-03  1.13e-01  1.96e-02]
   [-7.11e-03 -1.56e-04 -9.14e-02]
   [ 1.26e-01  8.60e-01  8.74e-02]
   [ 1.87e-01  3.62e-01  1.54e-01]
   [ 7.47e-02 -3.32e-01 -3.50e-01]
   [ 7.47e-02  1.61e-01 -1.08e-01]
   [ 3.05e-02  2.09e-01 -6.94e-02]
   [ 1.08e-01  5.49e-01 -4.98e-02]
   [-2.56e-01 -5.64e-01 -5.05e-02]
   [ 7.92e-02 -3.12e-01  8.78e-02]
   [-8.33e-02 -2.10e-01  2.24e-01]
   [-2.07e-01 -4.67e-01  9.76e-02]
   [-1.66e-02 -2.81e-01 -5.97e-02]
   [ 4.13e-02 -1.76e-01 -9.93e-02]
   [ 6.93e-02 -2.77e-01 -4.04e-01]
   [ 1.63e-01 -1.83e-01 -1.21e-01]
   [ 7.47e-02  8.95e-01  2.58e-03]
   [-2.49e-01 -4.29e-01 -1.11e-01]
   [ 3.11e-01  2.79e-01 -4.63e-02]
   [-1.29e-01  3.04e-01 -4.47e-02]
   [ 1.88e-01 -1.28e-01 -1.05e-01]
   [ 5.89e-02  6.53e-01  4.95e-02]
   [ 1.37e-01 -2.70e-01 -1.79e-01]
   [-2.32e-01 -4.30e-01  7.47e-02]
   [-3.62e-01 -3.57e-01  6.39e-02]
   [-1.04e-02  7.68e-01 -1.83e-02]
   [-7.95e-02 -6.00e-01 -3.28e-01]
   [ 8.28e-02  6.60e-02  6.64e-02]
   [-2.47e-02 -3.43e-01  2.91e-01]
   [ 2.49e-02  3.70e-01 -2.78e-01]
   [-1.87e-01  2.09e-02  2.61e-01]
   [ 1.66e-01  3.07e-01 -1.34e-01]
   [-1.68e-02  3.47e-01 -1.91e-04]
   [ 4.56e-02  8.92e-01 -2.91e-02]
   [ 1.05e-01  8.10e-01 -9.57e-02]
   [ 7.73e-02  3.66e-01 -2.25e-01]
   [-1.78e-01 -3.85e-01 -2.24e-01]
   [ 2.82e-02  6.06e-01 -2.39e-03]
   [ 0.00e+00 -2.99e-01 -8.24e-02]
   [ 5.81e-02 -4.54e-01 -2.54e-01]
   [ 8.30e-03  1.01e-01 -1.24e-01]
   [ 1.16e-01 -1.99e-01 -3.40e-02]
   [-1.16e-01 -5.90e-01 -5.62e-02]
   [-1.66e-02 -3.56e-01  9.13e-02]
   [-1.66e-02  8.11e-01 -1.12e-02]
   [-1.96e-01 -1.63e-01  1.03e-01]
   [-2.32e-01 -5.64e-01  3.32e-02]
   [ 8.74e-02  2.45e-01 -1.29e-01]
   [ 7.37e-02  3.74e-01  1.91e-01]
   [-1.18e-01 -3.33e-02 -2.43e-02]
   [-1.55e-01  3.59e-01  5.27e-02]
   [-2.27e-01 -2.41e-01  2.52e-01]
   [ 2.01e-02  1.40e-01  4.22e-02]
   [ 1.28e-01  8.30e-02  1.88e-01]
   [ 9.90e-02 -3.38e-01 -2.58e-01]
   [-1.18e-01 -4.40e-01  4.82e-02]
   [ 1.79e-01  2.12e-01  8.75e-02]
   [-1.43e-01  2.42e-01 -3.24e-02]
   [-6.19e-02 -1.53e-01  2.60e-01]
   [ 1.69e-02  2.55e-01 -1.32e-01]
   [-5.95e-02 -5.15e-01 -3.30e-01]
   [-1.42e-01 -4.31e-01  2.30e-01]
   [ 1.07e-02  5.81e-02 -8.13e-02]
   [-9.13e-02 -4.18e-01 -3.47e-01]
   [ 8.30e-02  8.35e-01 -1.08e-02]
   [-1.19e-01 -7.94e-02  1.59e-01]
   [-2.46e-01 -2.84e-01 -2.66e-01]
   [-1.40e-01 -5.87e-02 -5.77e-02]
   [-2.63e-01 -1.80e-02 -3.97e-02]
   [-1.47e-01 -5.48e-01 -2.82e-01]
   [ 1.36e-01 -1.24e-01  1.31e-01]
   [-9.66e-02 -2.11e-01  9.13e-02]
   [-5.02e-02  5.75e-02 -3.58e-02]
   [ 7.93e-02  2.77e-01 -4.71e-02]
   [ 2.86e-02  3.52e-01  1.03e-01]
   [ 4.15e-02  3.78e-01 -3.74e-02]]
# ---
# name: test_density_operator[True-False-False]
  []
# ---
# name: test_density_operator[True-False-True]
  []
# ---
# name: test_density_operator[True-True-False]
  [[0.78 0.91 1.54]
   [1.22 1.26 1.25]
   [1.26 0.84 0.78]
   [1.01 0.94 1.02]
   [0.85 0.71 0.93]
   [0.99 0.72 1.09]
   [1.19 1.48 1.23]
   [1.24 0.90 1.15]
   [1.07 0.92 1.62]
   [0.99 0.88 0.88]
   [1.31 0.89 0.82]
   [1.26 1.37 1.16]
   [1.01 0.99 1.56]
   [1.20 1.16 1.00]
   [1.40 1.38 1.16]
   [1.43 1.45 1.11]
   [0.80 0.67 1.27]
   [1.03 0.90 1.56]
   [1.21 0.90 1.06]
   [0.84 0.56 1.17]
   [1.28 1.38 0.99]
   [1.16 0.67 1.47]
   [0.81 0.67 1.09]
   [1.14 1.07 1.67]
   [1.16 1.44 1.06]
   [1.16 0.80 0.85]
   [1.12 1.38 1.25]
   [0.83 0.70 1.16]
   [1.05 0.97 1.43]
   [1.42 1.04 1.01]
   [0.81 0.80 1.30]
   [0.71 0.67 1.21]
   [1.02 1.01 1.30]
   [1.18 1.01 1.35]
   [1.25 1.29 1.13]
   [1.06 0.90 1.63]
   [1.28 1.45 1.28]
   [1.33 1.09 1.23]
   [1.06 0.70 1.07]
   [1.16 1.04 1.47]
   [1.11 0.98 1.06]
   [1.38 1.08 1.04]
   [1.08 0.79 1.18]
   [1.14 1.01 1.11]
   [0.89 0.71 1.20]
   [0.87 0.54 1.31]
   [1.23 1.83 1.38]
   [1.34 1.17 1.18]
   [1.02 1.05 1.13]
   [0.87 0.58 1.28]
   [1.24 0.96 1.03]
   [1.10 0.83 1.46]
   [1.07 0.95 1.32]
   [1.22 0.98 1.37]
   [1.22 1.23 1.11]
   [1.05 1.28 1.21]
   [0.91 1.00 1.35]
   [0.85 1.13 1.54]
   [1.28 1.88 1.17]
   [0.89 0.89 1.04]
   [1.01 0.93 1.11]
   [0.99 1.04 1.34]
   [1.00 0.72 1.01]
   [1.13 0.73 1.16]
   [1.07 0.79 1.35]
   [1.21 0.87 1.36]
   [1.24 0.96 1.07]
   [1.34 1.31 1.12]
   [1.28 0.87 1.31]
   [1.41 0.99 0.82]
   [0.93 0.81 1.22]
   [1.25 1.36 1.14]
   [0.73 0.80 1.37]
   [1.03 0.86 1.22]
   [1.33 1.16 1.03]
   [1.32 1.02 1.27]
   [0.82 0.78 1.10]
   [1.10 1.20 1.21]
   [1.16 0.59 0.80]
   [0.85 0.59 1.21]
   [1.09 1.44 1.17]
   [1.12 1.47 1.09]
   [1.24 1.00 1.37]
   [0.79 0.82 1.45]
   [1.29 1.52 1.20]
   [1.30 1.52 1.04]
   [1.17 1.59 1.18]
   [1.11 0.70 1.56]
   [1.16 1.30 1.01]
   [0.79 1.02 1.46]
   [1.06 0.91 1.41]
   [1.25 0.92 1.01]
   [1.09 1.80 1.29]
   [1.20 1.35 1.09]
   [1.03 0.49 1.27]
   [0.98 0.78 1.00]
   [0.98 0.92 0.92]
   [1.41 1.32 1.13]
   [0.89 0.75 1.29]
   [1.14 0.77 0.94]]
# ---
# name: test_density_operator[True-True-True]
  [[-0.41 -0.28  0.35]
   [ 0.02  0.06  0.05]
   [ 0.07 -0.35 -0.42]
   [-0.19 -0.26 -0.18]
   [-0.35 -0.49 -0.26]
   [-0.20 -0.47 -0.10]
   [-0.00  0.29  0.04]
   [ 0.05 -0.29 -0.05]
   [-0.13 -0.27  0.42]
   [-0.21 -0.31 -0.31]
   [ 0.12 -0.30 -0.37]
   [ 0.07  0.17 -0.03]
   [-0.19 -0.20  0.37]
   [ 0.01 -0.03 -0.20]
   [ 0.21  0.19 -0.03]
   [ 0.23  0.26 -0.08]
   [-0.39 -0.52  0.08]
   [-0.16 -0.30  0.37]
   [ 0.02 -0.30 -0.14]
   [-0.36 -0.64 -0.03]
   [ 0.09  0.18 -0.21]
   [-0.04 -0.52  0.27]
   [-0.39 -0.53 -0.10]
   [-0.05 -0.13  0.48]
   [-0.04  0.24 -0.13]
   [-0.03 -0.40 -0.34]
   [-0.07  0.19  0.06]
   [-0.37 -0.49 -0.04]
   [-0.15 -0.22  0.23]
   [ 0.23 -0.16 -0.18]
   [-0.38 -0.40  0.11]
   [-0.48 -0.53  0.01]
   [-0.17 -0.19  0.11]
   [-0.01 -0.19  0.15]
   [ 0.06  0.09 -0.06]
   [-0.13 -0.29  0.43]
   [ 0.08  0.26  0.08]
   [ 0.13 -0.11  0.04]
   [-0.14 -0.49 -0.13]
   [-0.03 -0.16  0.27]
   [-0.08 -0.22 -0.13]
   [ 0.19 -0.12 -0.15]
   [-0.12 -0.41 -0.02]
   [-0.06 -0.18 -0.09]
   [-0.31 -0.48  0.00]
   [-0.33 -0.66  0.12]
   [ 0.03  0.64  0.18]
   [ 0.14 -0.02 -0.02]
   [-0.17 -0.14 -0.06]
   [-0.32 -0.61  0.09]
   [ 0.05 -0.23 -0.16]
   [-0.09 -0.37  0.26]
   [-0.13 -0.24  0.13]
   [ 0.03 -0.21  0.17]
   [ 0.02  0.03 -0.09]
   [-0.15  0.08  0.01]
   [-0.28 -0.19  0.16]
   [-0.34 -0.07  0.34]
   [ 0.08  0.69 -0.03]
   [-0.31 -0.30 -0.16]
   [-0.18 -0.27 -0.09]
   [-0.20 -0.15  0.14]
   [-0.20 -0.47 -0.18]
   [-0.06 -0.46 -0.03]
   [-0.12 -0.40  0.15]
   [ 0.01 -0.32  0.17]
   [ 0.04 -0.23 -0.13]
   [ 0.14  0.12 -0.08]
   [ 0.09 -0.33  0.11]
   [ 0.22 -0.20 -0.38]
   [-0.27 -0.39  0.02]
   [ 0.06  0.17 -0.05]
   [-0.47 -0.39  0.17]
   [-0.17 -0.33  0.03]
   [ 0.14 -0.04 -0.16]
   [ 0.12 -0.18  0.07]
   [-0.38 -0.41 -0.10]
   [-0.09  0.01  0.02]
   [-0.04 -0.60 -0.39]
   [-0.34 -0.61  0.01]
   [-0.11  0.24 -0.03]
   [-0.08  0.28 -0.11]
   [ 0.05 -0.20  0.17]
   [-0.41 -0.37  0.26]
   [ 0.10  0.32  0.00]
   [ 0.10  0.32 -0.15]
   [-0.03  0.39 -0.01]
   [-0.09 -0.49  0.36]
   [-0.03  0.10 -0.18]
   [-0.40 -0.18  0.26]
   [-0.14 -0.29  0.22]
   [ 0.05 -0.27 -0.18]
   [-0.10  0.61  0.09]
   [ 0.01  0.16 -0.11]
   [-0.17 -0.71  0.07]
   [-0.21 -0.42 -0.20]
   [-0.22 -0.27 -0.27]
   [ 0.22  0.13 -0.06]
   [-0.30 -0.44  0.10]
   [-0.06 -0.42 -0.26]]
# ---
